components/org.wso2.carbon.identity.application.authenticator.push/src/main/java/org/wso2/carbon/identity/application/authenticator/push/PushAuthenticator.java

```diff
@@ -242,40 +254,164 @@ protected void processAuthenticationResponse(HttpServletRequest httpServletReque
     }
 
     /**
-     * Validate the signature using the JWT received from mobile app
+     * Validate the correlation between the challenged saved in the server and received from the auth response.
      *
-     * @param jwt       JWT generated from mobile app
+     * @param claimsSet JWT claim set for the validated auth response token
      * @param challenge Challenge stored in cache to correlate with JWT
-     * @return Boolean for validity of the signature
-     * @throws PushAuthenticatorException Exception in push authenticator
+     * @param deviceId  Unique ID for the device trying to authenticate the user
+     * @throws AuthenticationFailedException if the jwt fails to parse when getting the challenge claim
      */
-    private boolean validateSignature(String jwt, String challenge)
-            throws PushAuthenticatorException {
+    private void validateChallenge(JWTClaimsSet claimsSet, String challenge, String deviceId)
+            throws AuthenticationFailedException {
+
+        try {
+            if (claimsSet != null) {
+                if (!claimsSet.getStringClaim(PushAuthenticatorConstants.TOKEN_CHALLENGE).equals(challenge)) {
+                    String errorMessage = String
+                            .format("Authentication failed! Challenge received from %s  was not valid.", deviceId);
+                    throw new AuthenticationFailedException(errorMessage);
+                }
+            } else {
+                String errorMessage = String
+                        .format("Authentication failed! JWT claim set received from %s  was null.", deviceId);
+                throw new AuthenticationFailedException(errorMessage);
+            }
+        } catch (ParseException e) {
```

Comment:
Handle getting claims in a private method and throw authenticationFailedException